nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java

```diff
@@ -0,0 +1,376 @@
+    .required(false)
+    .expressionLanguageSupported(true)
+    .addValidator(Validator.VALID)
+    .build();
+
+  public static final PropertyDescriptor TIMEOUT = new PropertyDescriptor.Builder()
+    .name("Timeout")
+    .description("Timeout in milliseconds when writing events to Riemann")
+    .required(true)
+    .defaultValue("1000")
+    .addValidator(StandardValidators.POSITIVE_LONG_VALIDATOR)
+    .build();
+
+  private List<PropertyDescriptor> customAttributes = new ArrayList<>();
+  private static final Set<Relationship> relationships = new HashSet<>();
+  private static final List<PropertyDescriptor> localProperties = new ArrayList<>();
+
+  private int batchSize = -1;
+  private long writeTimeout = 1000;
+
+  static {
+    relationships.add(REL_SUCCESS);
+    relationships.add(REL_FAILURE);
+    localProperties.add(RIEMANN_HOST);
+    localProperties.add(RIEMANN_PORT);
+    localProperties.add(TRANSPORT_PROTOCOL);
+    localProperties.add(TIMEOUT);
+    localProperties.add(BATCH_SIZE);
+    localProperties.add(ATTR_DESCRIPTION);
+    localProperties.add(ATTR_SERVICE);
+    localProperties.add(ATTR_STATE);
+    localProperties.add(ATTR_METRIC);
+    localProperties.add(ATTR_TTL);
+    localProperties.add(ATTR_TAGS);
+    localProperties.add(ATTR_HOST);
+    localProperties.add(ATTR_TIME);
+  }
+
+
+  @Override
+  public Set<Relationship> getRelationships() {
+    return relationships;
+  }
+
+  @Override
+  protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+    return localProperties;
+  }
+
+  @Override
+  protected PropertyDescriptor getSupportedDynamicPropertyDescriptor(final String propertyDescriptorName) {
+    return new PropertyDescriptor.Builder()
+      .name(propertyDescriptorName)
+      .expressionLanguageSupported(true)
+      .addValidator(Validator.VALID)
+      .required(false)
+      .dynamic(true)
+      .build();
+  }
+
+  @OnStopped
+  public final void cleanUpClient() {
+    if (riemannClient != null) {
+      this.riemannClient.close();
+    }
+    this.riemannClient = null;
+    this.batchSize = -1;
+    this.customAttributes.clear();
+  }
+
+  @OnScheduled
+  public void onScheduled(ProcessContext context) throws ProcessException {
+    if (batchSize == -1) {
+      batchSize = context.getProperty(BATCH_SIZE).asInteger();
+    }
+    if (riemannClient == null || !riemannClient.isConnected()) {
+      transport = Transport.valueOf(context.getProperty(TRANSPORT_PROTOCOL).getValue());
+      String host = context.getProperty(RIEMANN_HOST).getValue().trim();
+      int port = context.getProperty(RIEMANN_PORT).asInteger();
+      writeTimeout = context.getProperty(TIMEOUT).asLong();
+      RiemannClient client = null;
+      try {
+        switch (transport) {
+          case TCP:
+            client = RiemannClient.tcp(host, port);
+            break;
+          case UDP:
+            client = RiemannClient.udp(host, port);
+            break;
+        }
+        client.connect();
```

Comment:
I believe this is the culprit in what works out to be a resource leak.  When testing against my sample instance, and the server went down (running in a docker container), I quickly ended up in an OOME.  

```
2015-11-04 11:31:19,111 ERROR [Timer-Driven Process Thread-1] o.a.nifi.processors.riemann.PutRiemann PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] Failed to process session due to org.apache.nifi.processor.exception.ProcessException: Unable to connect to Riemann [192.168.99.100:5555] (TCP)
Connection failed: org.apache.nifi.processor.exception.ProcessException: Unable to connect to Riemann [192.168.99.100:5555] (TCP)
Connection failed
2015-11-04 11:31:20,114 ERROR [Timer-Driven Process Thread-5] o.a.nifi.processors.riemann.PutRiemann PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] failed to process due to java.lang.OutOfMemoryError: unable to create new native thread; rolling back session: java.lang.OutOfMemoryError: unable to create new native thread
2015-11-04 11:31:20,114 ERROR [Timer-Driven Process Thread-5] o.a.nifi.processors.riemann.PutRiemann PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] failed to process session due to java.lang.OutOfMemoryError: unable to create new native thread: java.lang.OutOfMemoryError: unable to create new native thread
2015-11-04 11:31:20,114 WARN [Timer-Driven Process Thread-5] o.a.nifi.processors.riemann.PutRiemann PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] Processor Administratively Yielded for 1 sec due to processing failure
2015-11-04 11:31:20,115 WARN [Timer-Driven Process Thread-5] o.a.n.c.t.ContinuallyRunProcessorTask Administratively Yielding PutRiemann[id=db08f8c1-a7a1-41fc-96c4-36e350bbb44d] due to uncaught Exception: java.lang.OutOfMemoryError: unable to create new native thread
2015-11-04 11:31:20,119 WARN [Timer-Driven Process Thread-5] o.a.n.c.t.ContinuallyRunProcessorTask 
java.lang.OutOfMemoryError: unable to create new native thread
	at java.lang.Thread.start0(Native Method) [na:1.8.0_60]
	at java.lang.Thread.start(Thread.java:714) [na:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:950) [na:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1368) [na:1.8.0_60]
	at org.jboss.netty.util.internal.DeadLockProofWorker.start(DeadLockProofWorker.java:38) ~[na:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.openSelector(AbstractNioSelector.java:344) ~[na:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.<init>(AbstractNioSelector.java:96) ~[na:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.<init>(AbstractNioWorker.java:51) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioWorker.<init>(NioWorker.java:45) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioWorkerPool.createWorker(NioWorkerPool.java:45) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioWorkerPool.createWorker(NioWorkerPool.java:28) ~[na:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioWorkerPool.newWorker(AbstractNioWorkerPool.java:99) ~[na:na]
	at org.jboss.netty.channel.socket.nio.AbstractNioWorkerPool.init(AbstractNioWorkerPool.java:69) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioWorkerPool.<init>(NioWorkerPool.java:39) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioWorkerPool.<init>(NioWorkerPool.java:33) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioClientSocketChannelFactory.<init>(NioClientSocketChannelFactory.java:152) ~[na:na]
	at org.jboss.netty.channel.socket.nio.NioClientSocketChannelFactory.<init>(NioClientSocketChannelFactory.java:117) ~[na:na]
	at com.aphyr.riemann.client.TcpTransport.connect(TcpTransport.java:151) ~[na:na]
	at com.aphyr.riemann.client.RiemannClient.connect(RiemannClient.java:155) ~[na:na]
	at org.apache.nifi.processors.riemann.PutRiemann.onScheduled(PutRiemann.java:260) ~[na:na]
	at org.apache.nifi.processors.riemann.PutRiemann.onTrigger(PutRiemann.java:286) ~[na:na]
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27) ~[nifi-api-0.3.1-SNAPSHOT.jar:0.3.1-SNAPSHOT]
	at org.apache.nifi.controller.StandardProcessorNode.onTrigger(StandardProcessorNode.java:1134) ~[nifi-framework-core-0.3.1-SNAPSHOT.jar:0.3.1-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:127) [nifi-framework-core-0.3.1-SNAPSHOT.jar:0.3.1-SNAPSHOT]
	at org.apache.nifi.controller.tasks.ContinuallyRunProcessorTask.call(ContinuallyRunProcessorTask.java:49) [nifi-framework-core-0.3.1-SNAPSHOT.jar:0.3.1-SNAPSHOT]
	at org.apache.nifi.controller.scheduling.TimerDrivenSchedulingAgent$1.run(TimerDrivenSchedulingAgent.java:119) [nifi-framework-core-0.3.1-SNAPSHOT.jar:0.3.1-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_60]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) [na:1.8.0_60]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) [na:1.8.0_60]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) [na:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_60]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_60]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_60]
```

Performing a cursory look through the code, it seems like when that IOException occurs the associated client should also be closed